### ISSUE TITLE
GetFolder() method not getting all the fields.

### ISSUE THREAD

**Issue by anushrikinhikar** [jhoerr/box-csharp-sdk-v2#56]
Hi jhoerr,

I am calling the GetFolder() method like below:

boxManager.GetFolder(Folder.Root, new[] { FolderField.Etag, FolderField.Name });

but here it is giving me "Object reference not set to instance of an object" error.
When i debugged it, I found the error is coming in: 
line  while (folder.ItemCollection.Entries.Count < folder.ItemCollection.TotalCount)
in the below function since i m getting folder.ItemCollection = null

private Folder DoGetFolderSync(string id, IEnumerable<FolderField> fields, string etag, int? limit, int? offset, Func<BoxRestClient, BoxRestClient> prepareClient)
        {
            GuardFromNull(id, "id");
            var paginate = limit.HasValue || offset.HasValue;
            var request = _requestHelper.Get(ResourceType.Folder, id, fields, etag, paginate ? limit : MaxItems, paginate ? offset : 0);
            var folder = prepareClient(_restClient).ExecuteAndDeserialize<Folder>(request);

```
        if (!paginate)
        {
            offset = MaxItems;
            while (folder.ItemCollection.Entries.Count < folder.ItemCollection.TotalCount)
            {
                request = _requestHelper.Get(ResourceType.Folder, id, fields, etag, MaxItems, offset);
                var next = prepareClient(_restClient).ExecuteAndDeserialize<Folder>(request);
                folder.ItemCollection.Entries.AddRange(next.ItemCollection.Entries);
                offset += MaxItems;
            }
        }
        return folder;
    }
```

Please let me know what is the correct code to fetch all the details of the folder/file

**Comment by jhoerr** [comment]
Your code should work. Please post a Fiddler trace of the HTTP request and response that's generating this error.

**Comment by anushrikinhikar** [comment]
Hi John,

What I want is to access Box files under my account using .net and display
it in a web application.

I am using the V1 API method to create an object of BoxManager class.

 BoxManager boxManager = new BoxManager("MyAPIKey", "MyAuthToken");

Is it because of this that the function GetFolder is failing ?

It would be great if you could help me with the V2 BoxManager class.
I tried creating an instance by putting in the values "accesstoken",
"clientId", "clientSecret" and "refreshToken" as below:

_var authenticator = new TokenProvider(clientId, clientSecret);_
_var oAuthToken = authenticator.RefreshAccessToken(refreshToken);_
_accessToken = oAuthToken.AccessToken;_

_// Instantiate a BoxManager with your api key and a user's auth token_
_var boxManager = new BoxManager(accessToken);_

This works fine but what should be the step if the access token expires and
the refresh token expires.

Could you provide me some code for refreshing the access token from the
code since I dont want the user to experience any session timeout or
something like that.

Appreciate you response !!!

Thanks,
Anushri Kinhikar

**Comment by anushrikinhikar** [comment]
Hi John,

I dont have fiddler trace just the stack trace if this could help.

[BoxException]
   BoxApi.V2.BoxRestClientBase.Try(IRestRequest request,
HttpStatusCode lastResponse) +913
   BoxApi.V2.BoxRestClientBase.Execute(IRestRequest request) +66
   RestSharp.RestClient.Execute(IRestRequest request) +35
   BoxApi.V2.BoxRestClientBase.ExecuteAndDeserialize(IRestRequest request) +140
   BoxApi.V2.BoxManager.DoGetFolderSync(String id, IEnumerable`1
fields, String etag, Nullable`1 limit, Nullable`1 offset, Func`2
prepareClient) +590
   BoxApi.V2.BoxManager.GetFolder(String id, IEnumerable`1 fields,
String etag, Nullable`1 limit, Nullable`1 offset) +326
   WebApplication2._Default.Page_Load(Object sender, EventArgs e) in
C:\TestApp\BoxNetAPI\WebApplication2\WebApplication2\Default.aspx.cs:28
   System.Web.UI.Control.LoadRecursive() +71
   System.Web.UI.Page.ProcessRequestMain(Boolean
includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)
+3178

_I needed this on an urgent basis._

_Please let me know if you have any clue about this._

_Thanks,_

_Anushri Kinhikar_

**Comment by jhoerr** [comment]
Hi Anushri,

The V1 API has been deprecated by Box; you'll need to use the V2 version.

> ...what should be the step if the access token expires and the refresh token expires.

If the access token expires, simply call `authenticator.RefreshAccessToken(refreshToken)` to get a new access/refresh token pair. You then need to create a new instance of the BoxManager with the new access token.

The refresh token expires after 14 days. If it's allowed to expire the user must reauthenticate your application. You can work around this limitation by periodically refreshing access/refresh tokens that have been idle for several days.

> Could you provide me some code for refreshing the access token

You're already aware of and using the code to accomplish this task. The only recommendation I could make is to wrap your operations in a try/catch clause that looks for HTTP 401 errors. If a 401 is encountered, that will likely indicate that the access token is expired and needs to be refreshed. You will then need to refresh the token and retry the operation.

**Comment by anushrikinhikar** [comment]
Thanks John,

Any updates on the GetFolder() method.

Regards,
Anushri Kinhikar

**Comment by anushrikinhikar** [comment]
Hey John,

I need to use the refresh token as there in the screenshot right ?
This is the refresh token that expires 14 days ? Because this refresh token
is expiring for me after some minutes itself.

[image: Inline image 1]

**Comment by jhoerr** [comment]
> Any updates on the GetFolder() method.

As far as I know that method is working properly. I can't help you any more until you post a Fiddler trace of the HTTP request and response that's generating the error.

> Because this refresh token is expiring for me after some minutes itself.

I can't see the image you posted. But anyway, I don't think the refresh token is expiring after a few minutes -- that would effectively kill Box's entire API for all users. Are you aware that refresh token is single use? That is, when you use the refresh token, a new refresh token is issued along with the new access token?

**Comment by anushrikinhikar** [comment]
Hi John,

I will share the client_id and client_secret of my api application with
you.
Is it possible for you to just send me a working code snippet with the
connection to BoxManager class using my credentials and getting all folder
details in it.

Could you use my credentials and other keys to find this oauth2AccessToken
and the refresh token (which expires after 14 days) since the refresh token
that I am getting is valid only for some time. If I try to use the refresh
token after some time, it gives me error saying "invalid refresh_token"

client_id="****_"
client_secret="**_**"
username="anushri.kinhikar@gmail.com
pwd="****"
API application name= "ABCDEFGH"

Please try this and let me know.
I need this to be done asap.

Thanks,
Anushri Kinhikar

**Comment by jhoerr** [comment]
Anushri, I'm sorry, I can't do this for you right now. Have you looked at the [usage example](https://github.com/jhoerr/box-csharp-sdk-v2#usage-example)? It contains working code.

If this SDK isn't meeting your needs, you might check out Box's official [Windows SDK](https://github.com/box/box-windows-sdk-v2).

**Comment by anushrikinhikar** [comment]
Hi John,

I took exactly the same example code.
Ok, how do I find out the access token and refresh token ? Is there any way
of finding the refresh token for my api.
I used the link of Box Token Generator:
https://box-token-generator.herokuapp.com/
Here when i enter the client id and client secret, it gives me access token
and refresh token.
I use this access token for boxmanager class object. But after some time
this access token expires.
So, for refreshing this access token, I use the refresh token which i got
prior, then it gives me invalid refresh_token error.

Is there anything that I may be missing on. I am clueless and I need this
to do as soon as possible.

Thanks,
Anushri Kinhikar
